video/out/opengl/hwdec_ios.m

```diff
@@ -0,0 +1,295 @@
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <assert.h>
+
+#include <CoreVideo/CoreVideo.h>
+#include <OpenGLES/EAGL.h>
+#include <OpenGLES/ES2/gl.h>
+#include <OpenGLES/ES2/glext.h>
+
+#include "video/mp_image_pool.h"
+#include "hwdec.h"
+#include "common/global.h"
+#include "options/options.h"
+
+struct vt_gl_plane_format {
+    GLenum gl_format;
+    GLenum gl_type;
+    GLenum gl_internal_format;
+    char swizzle[5];
+};
+
+struct vt_format {
+    uint32_t cvpixfmt;
+    int imgfmt;
+    int planes;
+    struct vt_gl_plane_format gl[MP_MAX_PLANES];
+};
+
+struct priv {
+    struct mp_hwdec_ctx hwctx;
+    struct mp_vt_ctx vtctx;
+
+    CVPixelBufferRef pbuf;
+    CVOpenGLESTextureCacheRef gl_texture_cache;
+    CVOpenGLESTextureRef gl_planes[MP_MAX_PLANES];
+};
+
+static struct vt_format vt_formats[] = {
+    {
+        .cvpixfmt = kCVPixelFormatType_420YpCbCr8BiPlanarVideoRange,
+        .imgfmt = IMGFMT_NV12,
+        .planes = 2,
+        .gl = {
+            { GL_LUMINANCE, GL_UNSIGNED_BYTE, GL_LUMINANCE },
+            { GL_LUMINANCE_ALPHA, GL_UNSIGNED_BYTE, GL_LUMINANCE_ALPHA, "raaa" },
+        }
+    },
+};
+
+static struct vt_format *vt_get_gl_format(uint32_t cvpixfmt)
+{
+    for (int i = 0; i < MP_ARRAY_SIZE(vt_formats); i++) {
+        if (vt_formats[i].cvpixfmt == cvpixfmt)
+            return &vt_formats[i];
+    }
+    return NULL;
+}
+
+static struct vt_format *vt_get_gl_format_from_imgfmt(uint32_t imgfmt)
+{
+    for (int i = 0; i < MP_ARRAY_SIZE(vt_formats); i++) {
+        if (vt_formats[i].imgfmt == imgfmt)
+            return &vt_formats[i];
+    }
+    return NULL;
+}
+
+static struct mp_image *download_image(struct mp_hwdec_ctx *ctx,
+                                       struct mp_image *hw_image,
+                                       struct mp_image_pool *swpool)
+{
+    if (hw_image->imgfmt != IMGFMT_VIDEOTOOLBOX)
+        return NULL;
+
+    struct mp_image *image = NULL;
+    CVPixelBufferRef pbuf = (CVPixelBufferRef)hw_image->planes[3];
+    CVPixelBufferLockBaseAddress(pbuf, kCVPixelBufferLock_ReadOnly);
+    size_t width  = CVPixelBufferGetWidth(pbuf);
+    size_t height = CVPixelBufferGetHeight(pbuf);
+    uint32_t cvpixfmt = CVPixelBufferGetPixelFormatType(pbuf);
+    struct vt_format *f = vt_get_gl_format(cvpixfmt);
+    if (!f)
+        goto unlock;
+
+    struct mp_image img = {0};
+    mp_image_setfmt(&img, f->imgfmt);
+    mp_image_set_size(&img, width, height);
+
+    for (int i = 0; i < f->planes; i++) {
+        void *base    = CVPixelBufferGetBaseAddressOfPlane(pbuf, i);
+        size_t stride = CVPixelBufferGetBytesPerRowOfPlane(pbuf, i);
+        img.planes[i] = base;
+        img.stride[i] = stride;
+    }
+
+    mp_image_copy_attributes(&img, hw_image);
+
+    image = mp_image_pool_new_copy(swpool, &img);
+
+unlock:
+    CVPixelBufferUnlockBaseAddress(pbuf, kCVPixelBufferLock_ReadOnly);
+
+    return image;
+}
```

Comment:
That entire function seems duplicated from the OSX one. (Though ideally we'd implement AVHWFramesContext for VT in FFmpeg anyway, which would make this whole function go away.)